### Video Games Dataset: EDA
#### 1. Describe Dataset
- **Who:** The data was acquired from Kaggle and supplied by the user Gregory Smith (https://www.kaggle.com/gregorut/videogamesales). The data was scraped from www.vgchartz.com. 
- **What:** The dataset contains a list of video games with sales greater than 100,000 from 1980 to 2011. It contains information such as the platform the game was made available, year of release, genre, publisher, sales in NA, sales in JP, sales in EU, sales in the rest of the world and global sales (total). The data set also includes the rank of games in terms of overall sales. **NOTE: Sales are in millions**
- **When:** The data set was last updated 4 years ago but contains games released from 1980 to seemingly 2017. 
- **Why:** The video game industry is a very competitive yet profitable industry. While big companies with large amounts of resources have an edge over smaller companies, we have recently seen many small companies finding huge success. Not only in game creation but in the case of streamers for example, playing a game before it becomes mainstream might give you an edge against bigger name streamers. With this data set, we are able to gain insight into general idea such as performance of companies, most popular titles and genres. We are also able to dive deeper and look at changing genre popularities over time, regional preference in game genres/platforms, upcoming developer etc.  
- **How:** The data set was scraped from the www.vgzchartz.com website using BeautifulSoup. The scraping script can be found here (https://github.com/GregorUT/vgchartzScrape)

#### 2. Load Dataset

In [2]:
import pandas as pd
import numpy as np
import altair as alt
# from altair_saver import save
alt.renderers.enable('mimetype')
alt.data_transformers.enable('data_server')

DataTransformerRegistry.enable('data_server')

In [3]:
game = pd.read_csv("vgsales.csv")

#### 3. Explore Dataset

In [4]:
game.info()

print("\nPlatform:\n",game.Platform.unique(),"\nCount: ",game.Platform.nunique())
print("\nYear\n",game.Year.unique(),"\nCount: ",game.Year.nunique())
print("\nGenre\n",game.Genre.unique(),"\nCount: ",game.Genre.nunique())
print("\nPublishers\n",game.Publisher.unique()[0:15],"\nCount: ",game.Publisher.nunique())

print(game.sort_values("NA_Sales",ascending=False).head(5).iloc[:,0:6])
print(game.sort_values("EU_Sales",ascending=False).head(5).iloc[:,0:6])
print(game.sort_values("JP_Sales",ascending=False).head(5).iloc[:,0:6])
print(game.sort_values("Global_Sales",ascending=False).head(5).iloc[:,0:6])

print(game.groupby("Year").size())
print(game.groupby("Genre").size())
print(game.groupby("Platform").size())

nulls = game[game.isna().any(axis=1)]
nulls

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB

Platform:
 ['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' '3DS' 'PS4' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'GC' 'WiiU' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX'] 
Count:  31

Year
 [2006. 1985. 2008. 2009. 1996. 1989. 1984. 2005. 199

Rank                             Name Platform    Year         Genre  \
179      180                  Madden NFL 2004      PS2     NaN        Sports   
377      378                 FIFA Soccer 2004      PS2     NaN        Sports   
431      432       LEGO Batman: The Videogame      Wii     NaN        Action   
470      471       wwe Smackdown vs. Raw 2006      PS2     NaN      Fighting   
607      608                   Space Invaders     2600     NaN       Shooter   
...      ...                              ...      ...     ...           ...   
16427  16430                     Virtua Quest       GC     NaN  Role-Playing   
16493  16496                       The Smurfs      3DS     NaN        Action   
16494  16497  Legends of Oz: Dorothy's Return      3DS  2014.0        Puzzle   
16543  16546           Driving Simulator 2011       PC  2011.0        Racing   
16553  16556                   Bound By Flame     X360  2014.0  Role-Playing   

                                    Publisher  NA_Sales  EU_Sales  JP_Sales  \
179                           Electronic Arts      4.26      0.26      0.01   
377                           Electronic Arts      0.59      2.36      0.04   
431    Warner Bros. Interactive Entertainment      1.86      1.02      0.00   
470                                       NaN      1.57      1.02      0.00   
607                                     Atari      2.36      0.14      0.00   
...                                       ...       ...       ...       ...   
16427                                 Unknown      0.01      0.00      0.00   
16493                                 Unknown      0.00      0.01      0.00   
16494                                     NaN      0.00      0.01      0.00   
16543                                     NaN      0.00      0.01      0.00   
16553                                     NaN      0.00      0.01      0.00   

       Other_Sales  Global_Sales  
179           0.71          5.23  
377           0.51          3.49  
431           0.29          3.17  
470           0.41          3.00  
607           0.03          2.53  
...            ...           ...  
16427         0.00          0.01  
16493         0.00          0.01  
16494         0.00          0.01  
16543         0.00          0.01  
16553         0.00          0.01  

[307 rows x 11 columns]

#### 4. Initial thoughts?
- We have null values in Year and Publisher (Total of 307 rows with NaN values)
- Year is a float, we could probably turn it into an int to make it prettier
- We have 31 unique Platforms
- We have 39 unique years (one being NaN)
- We have 12 unique genres
- We have 578 unique publishers
- Looking at the top games in each region EU and NA are more similar in taste than JP
    -JP loves their Pokemon
    -Nintendo is the dominating Publisher in all regions 
- Years with the most releases are 2006 to 2011 with a peak of 1431 releases in 2009.

#### 5. Wrangling
The main concern would be the NaN values. In my oppinion, 

In [5]:
#game.Year = game.Year.astype("Int64")
print(nulls.groupby("Year").size())
print(nulls.groupby("Genre").size())
print(nulls.groupby("Platform").size())

Year
2004.0    19
2005.0     5
2007.0     1
2010.0     2
2011.0     3
2012.0     2
2014.0     2
2016.0     2
dtype: int64
Genre
Action          65
Adventure       12
Fighting        12
Misc            53
Platform        11
Puzzle          12
Racing          24
Role-Playing    18
Shooter         28
Simulation      19
Sports          42
Strategy        11
dtype: int64
Platform
2600    17
3DS     10
DS      32
GB       1
GBA     36
GC      14
N64      3
PC      22
PS       7
PS2     34
PS3     25
PSP     16
PSV      3
Wii     35
X360    31
XB      21
dtype: int64


#### 6. Research Questions/Visualization+Analysis

In [7]:
!pip install plotly

     |████████████████████████████████| 13.2 MB 506 kB/s eta 0:00:01     |█████████████████████████████▌  | 12.2 MB 503 kB/s eta 0:00:03
  Created wheel for retrying: filename=retrying-1.3.3-py3-none-any.whl size=11430 sha256=1e1586d54f5ab48caf5ff21d7ec2878853081d621216f20b5229505124adb03a
  Stored in directory: /home/aamir/.cache/pip/wheels/c4/a7/48/0a434133f6d56e878ca511c0e6c38326907c0792f67b476e56
Successfully built retrying


In [10]:
!pip install ipywidgets

     |████████████████████████████████| 121 kB 270 kB/s eta 0:00:01
     |████████████████████████████████| 2.2 MB 498 kB/s eta 0:00:01
     |████████████████████████████████| 243 kB 430 kB/s eta 0:00:01


In [13]:
len(game)

16598

In [14]:
import plotly as ply

fig = ply.graph_objects.FigureWidget()
fig.add_trace(fig.add_indicator(
    mode = "number",
    value = len(game)#,
#     title = {'text': "Games",'font': {'color': 'gold','size':20}},
#     number={'font':{'color': 'gold','size':50}},
#     domain = {'row': 0, 'column': 0}
))

ValueError: 
    Invalid element(s) received for the 'data' property of 
        Invalid elements include: [FigureWidget({
    'data': [{'mode': 'number', 'type': 'indicator', 'uid': '058dee4b-f25b-4056-9944-7c7c2359c199', 'value': 16598}],
    'layout': {'template': '...'}
})]

    The 'data' property is a tuple of trace instances
    that may be specified as:
      - A list or tuple of trace instances
        (e.g. [Scatter(...), Bar(...)])
      - A single trace instance
        (e.g. Scatter(...), Bar(...), etc.)
      - A list or tuple of dicts of string/value properties where:
        - The 'type' property specifies the trace type
            One of: ['area', 'bar', 'barpolar', 'box',
                     'candlestick', 'carpet', 'choropleth',
                     'choroplethmapbox', 'cone', 'contour',
                     'contourcarpet', 'densitymapbox', 'funnel',
                     'funnelarea', 'heatmap', 'heatmapgl',
                     'histogram', 'histogram2d',
                     'histogram2dcontour', 'image', 'indicator',
                     'isosurface', 'mesh3d', 'ohlc', 'parcats',
                     'parcoords', 'pie', 'pointcloud', 'sankey',
                     'scatter', 'scatter3d', 'scattercarpet',
                     'scattergeo', 'scattergl', 'scattermapbox',
                     'scatterpolar', 'scatterpolargl',
                     'scatterternary', 'splom', 'streamtube',
                     'sunburst', 'surface', 'table', 'treemap',
                     'violin', 'volume', 'waterfall']

        - All remaining properties are passed to the constructor of
          the specified trace type

        (e.g. [{'type': 'scatter', ...}, {'type': 'bar, ...}])

#### 7. Future Studies